### PR TITLE
Feat: Link Previews (OG Images & Video + oEmbed Videos & Tweets)

### PR DESCRIPTION
**Pull request by CamKem**
This pull request introduces the ability to preview links & embed tweets and videos, using a meta data service.

Current implementation will display a video for YouTube or Vimeo, a tweet for X.com using oEmbed api. Then if there is an og:image it will display a preview card. If there is non of these or the link is not valid, it will just display a regular anchor tag.

https://github.com/user-attachments/assets/74cebdb2-1f56-4df8-a212-f0abbd723559

Open to suggestions.

Generated list of changes:

### Metadata and Link Previews:
* [`app/Services/MetaData.php`](diffhunk://#diff-4c77a84861e54302647b49e257f9b551d725aad4b538d7dacd3ce635f7a92ea0R1-R157): Introduced a new `MetaData` service class to fetch and cache Open Graph and oEmbed data for a given URL. This includes methods to handle Twitter and YouTube oEmbed data.
* [`app/Services/ParsableContentProviders/LinkProviderParsable.php`](diffhunk://#diff-2a8623f5c9639c196b841eac1e1800643502b1ec25ce6273b982f53721db0f7eR8): Integrated the `MetaData` service to fetch metadata for URLs and render link preview cards if metadata is available. [[1]](diffhunk://#diff-2a8623f5c9639c196b841eac1e1800643502b1ec25ce6273b982f53721db0f7eR8) [[2]](diffhunk://#diff-2a8623f5c9639c196b841eac1e1800643502b1ec25ce6273b982f53721db0f7eR41-R55)
* [`resources/views/components/link-preview-card.blade.php`](diffhunk://#diff-dfc061f2aae22d299c02d41ced9e3faa95931fa739e1fdc53d53a804c0930477R1-R50): Added a new Blade component to render link preview cards based on fetched metadata.
* [`resources/views/livewire/questions/show.blade.php`](diffhunk://#diff-29daa964d7b5d7e4d9baec5ef0b4ac8277ad15e8721f4887850649f59e1934e9R321-R345): Enhanced the sharing functionality to handle content with link preview cards. [[1]](diffhunk://#diff-29daa964d7b5d7e4d9baec5ef0b4ac8277ad15e8721f4887850649f59e1934e9R321-R345) [[2]](diffhunk://#diff-29daa964d7b5d7e4d9baec5ef0b4ac8277ad15e8721f4887850649f59e1934e9R397)

### Theme Switcher Enhancements:
* [`resources/js/theme-switch.js`](diffhunk://#diff-b30018dee2db7bc2f521c0a3658c7082535cd09099e5c02fa966e8d09ad0cc59L2-R60): Improved the theme switcher to store and apply the current theme more effectively, including rendering tweets based on the theme.
* [`resources/views/layouts/components/head.blade.php`](diffhunk://#diff-fc5b53de1adc41a821d69ad843f8969f14c0eccbffd71a50ea638297c11e460fR210-R211): Added a script to load the Twitter widget asynchronously.

### Testing:
* [`tests/Unit/Services/MetaDataTest.php`](diffhunk://#diff-0ec8e05c5cd323dfa622c6ce8bd4e628e1861923bf75e844a8391bd7fb6d0c13R1-R61): Added unit tests for the `MetaData` service to ensure it correctly fetches and caches metadata, including handling failures.
* [`tests/Unit/Services/ContentTest.php`](diffhunk://#diff-221caf7c2118bedb3178206deb604aa8cbb3335fd2f4fdf71299552ae865a3a2R13-R51): Added tests to verify that only links with images or HTML oEmbeds are parsed and rendered correctly. [[1]](diffhunk://#diff-221caf7c2118bedb3178206deb604aa8cbb3335fd2f4fdf71299552ae865a3a2R13-R51) [[2]](diffhunk://#diff-221caf7c2118bedb3178206deb604aa8cbb3335fd2f4fdf71299552ae865a3a2L26-R93)

### CODE REVIEW

**Comment by CamKem** [comment]
Note: #633 should be merged to help with the loading time for link previews being parsed.

File: app/Services/MetaData.php [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+    {
+        $data = collect();
+
+        try {
+            $response = Http::get($this->url);
+
+            if ($response->ok()) {
+                $data = $this->parse($response->body())->filter(fn ($value): bool => $value !== '');
+            }
+        } catch (Exception) {
```
Comment (nunomaduro) [inline app/Services/MetaData.php:63]: Can you catch the specific exception you looking for?

File: tests/Unit/Services/ContentTest.php [hunk 2]
```diff
@@ -23,5 +62,33 @@
 
     $provider = new App\Services\ParsableContent();
 
-    expect($provider->parse($content))->toBe('<a data-navigate-ignore="true" class="text-blue-500 hover:underline hover:text-blue-700 cursor-pointer" target="_blank" href="https://pinkary.com/@nunomaduro">pinkary.com/@nunomaduro</a>');
+    expect($provider->parse($content))->toBe('<div
```
Comment (nunomaduro) [inline tests/Unit/Services/ContentTest.php:65]: You can use here `toMatchSnapshot`.

File: tests/Unit/Services/ContentTest.php [hunk 3]
```diff
@@ -10,6 +10,45 @@
     expect($provider->parse($content))->toBe('Sure, here is the link: <a data-navigate-ignore="true" class="text-blue-500 hover:underline hover:text-blue-700 cursor-pointer" target="_blank" href="https://example.com">example.com</a>. Let me know if you have any questions.');
 });
 
+test('only links with images or html oEmbeds are parsed', function () {
+    $content = 'Sure, here is the link: https://laravel.com. Let me know if you have any questions.';
+
+    $provider = new App\Services\ParsableContent();
+
+    expect($provider->parse($content))->toBe(
```
Comment (nunomaduro) [inline tests/Unit/Services/ContentTest.php:18]: `toMatchSnapshot`

File: resources/views/components/link-preview-card.blade.php [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+        <div data-tweet-id="{{ end($tweetId) }}" class="w-full overflow-hidden">
+            {!! $data->get('html') !!}
+        </div>
+    @else
+        <div class="w-full overflow-hidden">
+            {!! $data->get('html') !!}
+        </div>
+    @endif
+@elseif ($data->has('player'))
+    <iframe
```
Comment (nunomaduro) [inline resources/views/components/link-preview-card.blade.php:18]: is this totally safe?

File: resources/views/components/link-preview-card.blade.php [hunk 5]
```diff
@@ -0,0 +1,50 @@
+<div
+    id="link-preview-card"
+    data-url="{{ $url }}"
+    class="mx-auto mt-2 min-w-full group/preview" data-navigate-ignore="true"
+>
+@if ($data->has('html'))
+    @if ($data->get('site_name') === 'X (formerly Twitter)')
+        @php($tweetId = explode('/', $url))
+        <div data-tweet-id="{{ end($tweetId) }}" class="w-full overflow-hidden">
+            {!! $data->get('html') !!}
```
Comment (nunomaduro) [inline resources/views/components/link-preview-card.blade.php:10]: Is this safe?

File: resources/views/components/link-preview-card.blade.php [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+        </div>
+    @else
+        <div class="w-full overflow-hidden">
+            {!! $data->get('html') !!}
```
Comment (nunomaduro) [inline resources/views/components/link-preview-card.blade.php:14]: Is this safe?

File: app/Services/MetaData.php [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+    {
+        return (new self($url))->data;
+    }
+
+    /**
+     * Get the meta-data for a given URL.
+     *
+     * @return Collection<string, string>
+     */
+    public function getData(): Collection
```
Comment (nunomaduro) [inline app/Services/MetaData.php:53]: this becomes static too.

File: app/Services/MetaData.php [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+     */
+    private Collection $data;
+
+    /**
+     * Fetch the Open Graph data for a given URL.
+     */
+    public function __construct(private string $url)
+    {
+        /** @var Collection<string, string> $cachedData */
+        $cachedData = Cache::remember(
```
Comment (nunomaduro) [inline app/Services/MetaData.php:29]: can you lazy evaluate this? so basically the code looks like this:
```

$metaData = new MetaData('my-url');
$metaData->fetch(); // calls are done here...

File: app/Services/MetaData.php [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+                    ->each(function ($attribute) use ($data, $meta): void {
+                        $key = explode(':', $attribute)[1];
+                        if (! $data->has($key)) {
+                            $data->put($key, $meta->getAttribute('content'));
+                        }
+                    });
+            }
+        }
+
+        // fetch oEmbed data for X / Twitter
```
Comment (nunomaduro) [inline app/Services/MetaData.php:135]: i think you remove remove docs like this...

**Review by nunomaduro — CHANGES_REQUESTED** [review]

File: app/Services/MetaData.php [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+     * Get the meta-data for a given URL.
+     *
+     * @return Collection<string, string>
+     */
+    public function getData(): Collection
+    {
+        $data = collect();
+
+        try {
+            $response = Http::get($this->url);
```
Comment (MrPunyapal) [inline app/Services/MetaData.php:58]: What about Http::head 🤔

File: app/Services/MetaData.php (hunk 9 above)
Comment (CamKem) [reply app/Services/MetaData.php:135]: Done.

File: app/Services/MetaData.php (hunk 8 above)
Comment (CamKem) [reply app/Services/MetaData.php]: Done, however needed to either store the value of the fetch calls in a seperate variable, so we can then check if there collection is empty, etc, or store have fetch update the data property, then have a seperate getter for it.

Wasn't sure which way you'd prefer, so I went with the simplest, however can change easily.

```php
$service = new MetaData($url);
$metadata = $service->fetch();
```

File: resources/views/components/link-preview-card.blade.php (hunk 4 above)
Comment (CamKem) [reply resources/views/components/link-preview-card.blade.php]: It should be safe, but I have removed it for now, in favour of the oEmbed players, as that gives us a way to ensure html is secure by having access control over which oEmbed apis are called.

File: resources/views/components/link-preview-card.blade.php (hunk 5 above)
Comment (CamKem) [reply resources/views/components/link-preview-card.blade.php:10]: It's better if we prevent any bad actors from being able to add a og:html tag that will be parsed. So the only html keys will be from oEmbed calls from which we controller the source.

It's fixed by adding a list of allowed tags:
```php
$allowed = ['title', 'description', 'keywords', 'image', 'site_name', 'url', 'type'];
```

Then only adding the tag to the collection if its exists in the allowed array.
```php
  collect(['name', 'property'])
      ->map(fn ($name): string => $meta->getAttribute($name))
      ->filter(fn ($attribute): bool => in_array(explode(':', (string) $attribute)[0], $interested_in))
      ->each(function ($attribute) use ($data, $allowed, $meta): void {
          $key = explode(':', $attribute)[1];
          if (! $data->has($key) && in_array($key, $allowed, true)) {
              $data->put($key, $meta->getAttribute('content'));
          }
      });
```

File: resources/views/components/link-preview-card.blade.php (hunk 6 above)
Comment (CamKem) [reply resources/views/components/link-preview-card.blade.php:14]: See above comment, has been made more safe.

File: tests/Unit/Services/ContentTest.php (hunk 2 above)
Comment (CamKem) [reply tests/Unit/Services/ContentTest.php]: Done.

File: resources/views/components/link-preview-card.blade.php (hunk 4 above)
Comment (CamKem) [reply resources/views/components/link-preview-card.blade.php]: Done

File: tests/Unit/Services/ContentTest.php (hunk 3 above)
Comment (CamKem) [reply tests/Unit/Services/ContentTest.php]: Done

File: app/Services/MetaData.php (hunk 1 above)
Comment (CamKem) [reply app/Services/MetaData.php]: Changed to specific exception - `ConnectionException`

File: app/Services/MetaData.php (hunk 7 above)
Comment (CamKem) [reply app/Services/MetaData.php]: Should it be static? It needs to call $this->url & $this->parse($response).
Unless you were meaning to change the method in a different context?
Happy to do whatever you recommend.

File: app/Services/MetaData.php (hunk 10 above)
Comment (CamKem) [reply app/Services/MetaData.php:51]: Interesting though, but we need the <head> tag from within the response body, Http::head returns the request headers from the url.

**Pull request merged**